lib/mongos.js

```diff
@@ -261,17 +261,20 @@ class Mongos extends EventEmitter {
             options: self.options
           });
 
-          //
           // Only emit event at start
           if (self.state === 'stopped') {
             if (stdout.indexOf('waiting for connections') !== -1) {
-              waitForAvailable(self.options.bind_ip, self.options.port, err => {
-                if (err) return reject(err);
+              co(function*() {
```

Comment:
This looks wrong. This is a `Promise` wrapping an instance of `co` above, so we're already a few layers deep into confusion at this point. My guess is that you can probably just call `yield` here without calling `co` again, but this is also a case where I might simply not use `co`, and directly use `waitForAvailable` directly as a `Promise`. Remember that you are calling this from the callback for an event emitter, so your scoping is rather complicated

Reply:
Hmm, now I'm reading this more closely below (this same issue happens in `server`). Was the only motivation for this change to use `co`? In this case I think it's actually hurting readability to have nested `co` invocations. If we're not able to use the top-level `co` call, then I'm not sure this change makes a ton of sense.

Reply:
I actually wasn't sure if the nested `co`'s are needed. This was a question I had for you? I can try to do more research into it and see if it will work without.

I understand that the nested `co` is not optimal and I agree it is ugly, but hopefully my other PR will be accepted (or some other fix) that makes it irrelevant. However, I understand that each PR should be evaluated on it's own without assuming a different PR will fix it.

I actually wasn't sure if the nested `co`'s are needed.

Reply:
Your other PR uses this in the same way, but does not introduce additional `co` invocations. I would say yes, it should be investigated whether you need nested `co` calls in these cases. I would suspect that you do not.